### PR TITLE
audit all contributions by @devnexen for correctness and delete unnecessary libc ABI bits

### PR DESCRIPTION
**Pull request by devnexen**

### CODE REVIEW

**Comment by andrewrk** [comment]
@mikdusan is there anything here that you want to keep? If so, would you be willing to audit it for correctness, and then *add* a commit to this branch to keep it?

I will do the same.

**Comment by rofrol** [comment]
What happend here?

**Comment by andrewrk** [comment]
Context: https://github.com/ziglang/zig/pull/16333#issuecomment-1625920518

**Comment by lin72h** [comment]
@devnexen Thanks for working on these freebsd integration bits, I've been following all your PR and learned a lots about freebsd internal.

**Comment by likern** [comment]
Why there isnt any integration tests written to run under BSD running in QEMU? I think that would solve the problem with correctness and trust, isn't it?

**Comment by mikdusan** [comment]
> Why there isnt any integration tests written to run under BSD running in QEMU? I think that would solve the problem with correctness and trust, isn't it?

- bsd-user is not supported on a linux host
- full disclosure, I have never tried bsd-user on any system, but it looks like it needs to run on a bsd host
- emulating a bsd system is really the only way we are going to get a bsd host and is something we'd like to do and there appears to be a path to that goal but it just hasn't happened yet

IMO, we have a few issues about libc on BSD. It is not very well organized and lacks direct tests. Basically libc only gets test coverage to the extent zig's std depends on it. Some further thoughts; feel free to critique:

1. A direct libc test suite. We could create a test for freebsd `mmap()` special `MAP_ALIGNED(n)` flag and verify the resulting alignment. Should be a simple test and yet it verifies that we are using the correct binary flags. It also doesn't depend on zig std somehow requiring to use it.

2. It is possible that a direct libc test suite could get "out of hand" to be bundled with zig. If that is or becomes the case, there is nothing preventing us from launching a libc testsuite repo for zig.

3. libc tests optioned to indicate host capabilities for testing; eg: `-Dnetwork-ipv6`,`-Dnetwork-ipv4`, `-Dnetwork-internet`, `-Dnetwork-dns`, `-Dnetwork-icmp`. If a host doesn't have the capability, skip.

File: lib/std/c/openbsd.zig [hunk 1]
```diff
@@ -1137,18 +1127,6 @@ pub usingnamespace switch (builtin.cpu.arch) {
             sc_cookie: c_long,
         };
     },
-    .aarch64 => struct {
```
Comment (semarie) [inline lib/std/c/openbsd.zig:1140]: the struct `ucontext_t` is valid for openbsd aarch64, and useful for segfault handling

File: lib/std/c/openbsd.zig [hunk 2]
```diff
@@ -1615,39 +1593,3 @@ pub const KERN = struct {
...
-    pub const VERSION = 16;
-    pub const SERIALNO = 17;
-    pub const UUID = 18;
-    pub const PHYSMEM64 = 19;
-    pub const USERMEM64 = 20;
-    pub const NCPUFOUND = 21;
-    pub const ALLOWPOWERDOWN = 22;
-    pub const PERFPOLICY = 23;
-    pub const SMT = 24;
-    pub const NCPUONLINE = 25;
```
Comment (semarie) [inline lib/std/c/openbsd.zig:1644]: `NCPUONLINE` is required for `getCpuCount()` (in Thread)

File: lib/std/c/openbsd.zig [hunk 3]
```diff
@@ -1615,39 +1593,3 @@ pub const KERN = struct {
...
-    pub const POWER = 26;
-};
-
-/// TODO refines if necessary
-pub const PTHREAD_STACK_MIN = switch (builtin.cpu.arch) {
```
Comment (semarie) [inline lib/std/c/openbsd.zig:1649]: `PTHREAD_STACK_MIN` is correct

File: lib/std/Thread.zig [hunk 4]
```diff
@@ -695,7 +695,7 @@ const PosixThreadImpl = struct {
         defer assert(c.pthread_attr_destroy(&attr) == .SUCCESS);
 
         // Use the same set of parameters used by the libc-less impl.
-        const stack_size = @max(config.stack_size, c.PTHREAD_STACK_MIN);
```
Comment (semarie) [inline lib/std/Thread.zig:698]: `PTHREAD_STACK_MIN` usage make sense. `pthread_attr_setstacksize()` would return `EINVAL` if the value passed is less than `PTHREAD_STACK_MIN`

File: lib/std/c/openbsd.zig [hunk 5]
```diff
@@ -251,48 +251,48 @@ pub const EAI_MAX = 15;
 
 pub const msghdr = extern struct {
```
Comment (semarie) [inline lib/std/c/openbsd.zig:252]: the rename of the fields aligns the structure with linux names.

File: lib/std/c/openbsd.zig [hunk 6]
```diff
@@ -1615,39 +1593,3 @@ pub const KERN = struct {
...
-    pub const DISKCOUNT = 10;
-    pub const SENSORS = 11;
-    pub const CPUSPEED = 12;
-    pub const SETPERF = 13;
-    pub const VENDOR = 14;
-    pub const PRODUCT = 15;
-    pub const VERSION = 16;
-    pub const SERIALNO = 17;
-    pub const UUID = 18;
-    pub const PHYSMEM64 = 19;
```
Comment (semarie) [inline lib/std/c/openbsd.zig:1638]: `PHYSMEM64` is used by `totalSystemMemory()` (in lib/std/process.zig)

File: lib/std/process.zig [hunk 7]
```diff
@@ -1163,37 +1163,6 @@ pub fn totalSystemMemory() TotalSystemMemoryError!usize {
...
-                .netbsd => "hw.physmem64",
-                else => "hw.physmem",
-            };
-            os.sysctlbynameZ(name, &physmem, &len, null, 0) catch |err| switch (err) {
-                error.NameTooLong, error.UnknownName => unreachable,
-                else => return error.UnknownTotalSystemMemory,
-            };
-            return @as(usize, @intCast(physmem));
-        },
-        .openbsd => {
```
Comment (semarie) [inline lib/std/process.zig:1180]: the implementation for OpenBSD makes sense.

**Review by semarie — COMMENTED** [review]
I added lines comments regarding OpenBSD contributions for what I am sure it is need or important to have. I didn't comment for others BSDs.

**Comment by likern** [comment]
> 

I thought of more like running **OpenBSD over full virtualization** using `kvm`, which seems supports it according to [this](https://www.linux-kvm.org/page/Guest_Support_Status#UNIX_Family:_BSD), at least having it run and failed on PR and commits.

To run it on GitHub Actions seems it need to have _nested virtualization_, `kvm` inside GA. It's **not supported** as is [*BSD natively](https://github.com/actions/runner/issues/385).

**BUT** `macos` machines **does support this** (can run Virtualbox). There is [GitHub Action](https://github.com/vmactions/openbsd-vm) which supports it using this method.

_Another way_ is to run **self-hosted runners on native BSD system** and run GA on VM (containers?) using `bhyve`, as shown here https://github.com/myb-project/guide/blob/main/en/gh_runners.md. It allows to run _DragonflyBSD, FreeBSD,  NetBSD, OpenBSD_.

Similar approach is to run **self-hosted runner on Linux** with `kvm` support on it and run VMs on it.

File: lib/std/Thread.zig (hunk 4 above)
Comment (andrewrk) [reply lib/std/Thread.zig:698]: Unfortunately the values used for most of the operating systems are not verified, so this commit will remain reverted. In practice, 100% of the time `config.stack_size` will be orders of magnitude larger than the pthread stack minimum size.

**Comment by andrewrk** [comment]
I re-did the work myself over here, in order to avoid unnecessary trust: #16636

Thank you for your comments @semarie - I will go over them one by one now.

**Comment by andrewrk** [comment]
To all: testing is not a solution to sloppy contributions. Testing is certainly an important component of quality assurance, but when it comes to external ABIs, there is no substitution for expertise and carefulness. There could easily be extern functions that pass automated tests, yet expose a subtle flaw that could be exploited or triggered under certain conditions, and all of these commits are exactly the kind that are in danger of creating that exact scenario.

File: lib/std/c/openbsd.zig (hunk 2 above)
Comment (andrewrk) [reply lib/std/c/openbsd.zig:1644]: This was reverted by @devnexen incorrectly and is not reverted in #16636.

File: lib/std/c/openbsd.zig (hunk 6 above)
Comment (andrewrk) [reply lib/std/c/openbsd.zig:1638]: Incorrectly reverted by @devnexen in this PR; not reverted in #16636.

File: lib/std/process.zig (hunk 7 above)
Comment (andrewrk) [reply lib/std/process.zig:1180]: incorrectly reverted by @devnexen; not reverted in #16636

**Pull request closed without merging**